src/base/amo.hh

```diff
@@ -126,13 +126,19 @@ class AtomicGenericPair3Op : public TypedAtomicOpFunctor<T>
     std::function<void(T*, std::array<T, 2>&, std::array<T, 2>)> op;
 };
 
+// IMPORTANT: The following operations assume the guest is little-endian.
+
 template<typename T>
 class AtomicOpAnd : public TypedAtomicOpFunctor<T>
 {
     // Bitwise operations are only legal on integral types
     template<typename B>
     typename std::enable_if<std::is_integral<B>::value, void>::type
-    executeImpl(B *b) { *b &= a; }
+    executeImpl(B *b) {
+        B mem_data = letoh(*b);
```

Comment:
maybe I'm missing something obvious, but since a given ISA is (in theory) using the same endianness, wouldn't we want letoh or htole but not both?

Reply:
We first have to read the data from memory (which is in *guest* endianness, LE specifically in the case of the GPU) and convert it to the host endianness. Then, the data is loaded in the host register, and the host does the operation. After the host has the answer (in host endianness), we then need to write the data back in to guest memory *in guest endianness*. So, there's another conversion from host to LE (because we're assuming the guest is LE)